### ISSUE TITLE
iOS App: "Copy" button in "Home" tab doesn't work

### ISSUE THREAD

**Issue by nikslor** [CollaboraOnline/online#5839]
**Describe the bug**
"Copy" button in the "Home" tab of the current TestFligh verson (22.05.202302092239) is not working.

**To Reproduce**
1. Open a writer document in the TestFlight iOS app
2. Write some text and mark it
3. Tap the "Copy" button in the "Home" tab

**Expected behavior**
The "Paste" button should change its state and become "active" so the user can paste the copied content.

**Actual behavior**
The "Paste" button doesn't change its state and also the keyboard shortcut "cmd + v" will not work - so presumably the "Copy" function didn't work.

**Screenshots**
![image](https://user-images.githubusercontent.com/1211300/218397634-ee1ae1b7-7a9c-4bc6-9441-0e18f63e88c4.png)

**Tablet**
 - OS: iPadOS 15.6.1
 - Version: 22.05.202302092239

**Comment by Ezinnem** [comment]
I tested this and I could not reproduce on my end.

@nikslor , please can you test again?

**Comment by nikslor** [comment]
@Ezinnem it's indeed better now; but still not perfectly good. I sometimes still encounter the situation that the "Paste" button in the menu bar is no clickable (after using the "Copy" button) - exactly how you describe it in #5908. But I could not yet find the pattern that triggers this. FYI: I also encountered it in "writer". I'll test this a bit more and think we should not yet close the issue.

**Comment by plubius** [comment]
Although this is related to issue #5908, this bug is definitely its own separate bug.

From my testing, it appears that the enabled/disabled state of the Paste button is set only once at when the document is opened. So if the system clipboard is empty, the Paste button is disabled.

I will investigate if the something on either the iOS side is failing to notify the JS code that the state of the system clipboard has changed.

**Comment by mmeeks** [comment]
Hmm - so - I'm rather surprised that paste is ever inactive. Ultimately we can't easily tell if there is any content on the system clipboard from the browser - so I'd expect Paste to be constantly enabled ~everywhere.

**Comment by plubius** [comment]
> Hmm - so - I'm rather surprised that paste is ever inactive. Ultimately we can't easily tell if there is any content on the system clipboard from the browser - so I'd expect Paste to be constantly enabled ~everywhere.

I haven't looked that closely at the JS code but it does seem that the JS code is polling the clipboard state when a document is opened.

I cleared my Mac's general pasteboard and opened up a Writer and a Calc document in Collabora Online running in Safari. It appears to me that Online does not poll the clipboard and the Edit > Paste menu item is always enabled.

Anyway, I will investigate further and see if I can determine where the Online code is disabling pasting so that pasting is always enabled.

**Comment by plubius** [comment]
After a little debugging, I found that it is the LibreOffice code that is disabling the Paste toolbar button. The LibreOffice code is able to detect if the clipboard is empty and then sends the following message to the JS code:

[ app2js ] TRC  To JS: statechanged: .uno:Paste=disabled| CODocument.mm:99

I will see if I can ignore that message in the JS code if we are running in the iOS app since such a message won't be sent when running in a web browser. Hopefully that will prevent disabling of pasting.

**Comment by plubius** [comment]
My fix is in the latest test patch. Can you close the bug if the fix works for you? Feel free to keep it open if you find any unexpected behavior.

**Comment by nikslor** [comment]
It works for me and the bug seems to be fixed. What I find misleading is, that both, the copy and the paste button stay active after I taped on it. I tend to say, that the button should only have an active state for a blink of a moment and then go back into the non-active state.

But this is probably something I should add as a new issue. So let's close this one (as copy/paste now works).

Thanks @plubius!

**Comment by mmeeks** [comment]
@vmiklos one to watch. I worry that unless something has been copied in a COOL/CODE Kit process first - the core will think the (remote OS clipboard to which it has noaccess/visibility) is content-free, so by disabling this we make it impossible to paste into COOL from the OS - until we copy something inside COOL to make it believe there is something on the clipboard =)

**Comment by vmiklos** [comment]
Yes, I see the concern. I tested this scenario manually, and seems to work. (So the browser and the iOS case should be OK, I'm unsure about Android.) By accident, our paste tests always executes this scenario, too. I'll add a test to also make sure that in case we have something on the clipboard and we did an internal copy first, then the real clipboard is preferred over our internal clipboard (unless we find a marker and it's our marker).
